csrc/transformer/inference/csrc/apply_rotary_pos_emb.cu

```diff
@@ -116,10 +108,11 @@ __global__ void apply_rotary_pos_half(T* mixed_query,
         LAUNCH_ROT_POS_EMB_HALF(16, ALIGNMENT); \
     } else if (threads_per_head == 32) {        \
         LAUNCH_ROT_POS_EMB_HALF(32, ALIGNMENT); \
+    } else if (threads_per_head == 64) {        \
+        LAUNCH_ROT_POS_EMB_HALF(64, ALIGNMENT); \
```

Comment:
@jagadish-amd, are you confident this logic will work on non-amd since it only previously ran on amd? Would it be safer to condition this to amd?

Reply:
Actually that brings up a good question: what would be the right set of deepspeed unit tests or workloads to test the various cases in this logic block?

Reply:
there are several unit tests here for the different operations we support: https://github.com/microsoft/DeepSpeed/tree/master/tests/unit/ops/transformer/inference. The RoPE one is missing though :) I think it would be nice to add one, @jagadish-amd wound you mind adding one to test these different dimensions? Thanks

Reply:
@jagadish-amd, as Tunji mentioned too this new condition is not valid in Nvidia GPUs, as the Warp size is 32 vs the 64 for AMD's wavefront. So, the kind of blocks we are creating [here](https://github.com/microsoft/DeepSpeed/blob/master/csrc/transformer/inference/csrc/apply_rotary_pos_emb.cu#L42) can result into two warps and thus we cannot safely run the shfl instruction that is needed for [this part](https://github.com/microsoft/DeepSpeed/blob/master/csrc/transformer/inference/csrc/apply_rotary_pos_emb.cu#L75). I think we have a better implementation for this here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/inference/v2/kernels/ragged_ops/linear_blocked_kv_rotary/blocked_kv_rotary_cuda.cu which reads the data in larger granularity which ensures the warp-size is not exceeded here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/inference/v2/kernels/ragged_ops/linear_blocked_kv_rotary/blocked_kv_rotary_cuda.cu#L49

Reply:
Thanks all for the inputs. I added code only for warpsize 64 and retained old code for warpsize 32.